### ISSUE TITLE
Alt+Tab on FullScreen Direct3D game hangs on 3.6.0.X

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#1287]
**Describe the bug**
This bug has been reported on the forums as crashing here: https://www.adventuregamestudio.co.uk/forums/index.php?topic=58976.msg636635407#msg636635407

I actually don't have a crash on my Windows 10 PC, instead when I alt+tab a fullscreen Direct3D game, it just hangs with the window staying in the background, but I can't alt+tab back to it and the only way to close it is by killing using the task manager

**AGS Version**
Built directly from the master branch https://github.com/adventuregamestudio/ags/tree/3d0353e36ef55b14fffe3f7515d74adf095415a1

**Game**
Here's an example game, but it happens with any on this engine build (3.6.0.4 release happens too) [game.zip](https://github.com/adventuregamestudio/ags/files/6525699/Release.zip)

**To Reproduce**
- Make sure the Direct3D is in use (check with Ctrl+V). If not, set using winsetup.
- make sure the game is fullscreen (Alt+Enter if needed)
- then, alt+tab 
- the game will hang in the background, you can't alt+tab back to it's window

**Expected behavior**
It's expected to alt+tab to any other window on the desktop and alt+tab back

**Desktop:**
 - OS: Windows
 - Version 10

**Comment by ivan-mogilko** [comment]
I get a different error when running on Windows 7:

> Error: Unable to lock texture

Good thing that the problem is reproducible in some way.

**Comment by ivan-mogilko** [comment]
Ah. AGS does not stop running the game and keep redrawing even when window has no focus. It could be that suspending the alt+tabbed game is not working in SDL version yet.

Maybe Direct3D does not like that in fullscreen mode. But I don't remember how it worked in previous versions when Multitasking mode was on. Might be worth comparing, maybe there was some condition that prevented renderer from working that is now not active.

**Comment by ericoporto** [comment]
> AGS does not stop running the game and keep redrawing even when window has no focus

Uhm. AGS has `System.HasInputFocus` in the script API to be able to detect if the game is focused or not, does this means it should keep running just not drawing?

**Comment by ivan-mogilko** [comment]
AGS has two modes: multitasking and non-multitasking, they are controlled by [SetMultitaskingMode](https://github.com/adventuregamestudio/ags-manual/wiki/Globalfunctions_General#setmultitaskingmode). In multitasking mode engine supposed to run game loops when alt+tabbed, otherwise it should suspend game execution.
(Default is non-multitasking)

I keep forgetting details about allegro version, allegro probably had this as a library feature which AGS relied upon. This is currently missing in SDL version.

As for drawing, I think that this may be dependent on a gfx mode and maybe on renderer too. For example game may loose focus but still be on screen in windowed mode. Or loose focus but be fullscreen on another display (don't remember if it works like that, but probably does?). So this is a complex condition.
In case of Direct3D - it seems it does not like when it is in exclusive (fullscreen) mode, minimized (after alt-tabbing) and getting certain calls.

**Comment by ericoporto** [comment]
Erh, in the manual page you linked there's the following comment:

> Note that mode 1 does not work with some graphics cards in full-screen mode, so you should only rely on it working when your game is run in windowed mode.

This kinda sounds familiar.

**Comment by ivan-mogilko** [comment]
So, more bad news, apparently the game is no longer suspends upon loosing focus since AGS 3.5.1. It works in 3.5.0.
I think this broke after Nick Sonneveld changed the update mechanic from using allegro's timer to the time period calculation + sleep in the engine itself. My hypothesis is that allegro pauses its timers when told to suspend, but the new mechanic does not handle this.
So now this has to be fixed both in 3.5.1 and 3.6.0.

**Comment by ericoporto** [comment]
SDL has a timer of it's own, does using it helps in any way?

**Comment by ivan-mogilko** [comment]
It's not about having a timer, it's about being able to suspend it anytime.

**Comment by ivan-mogilko** [comment]
It's not about having a timer, it's about being able to suspend it anytime. Or, alternatively, keep running in a loop until certain flag is set/removed. SDL is only useful here if it provides utilities, but 3.5.1 does not have SDL anyway.

**Comment by ivan-mogilko** [comment]
I made a pr for the suspension issue (#1292), but unfortunately that won't be enough to fix Direct3D problem in SDL2 version because there's more in it; tried to explain in a [comment](https://github.com/adventuregamestudio/ags/pull/1292#issuecomment-846560058).

**Comment by ericoporto** [comment]
The change in #1288 will prevent the crash on Directx, and it's a single line change, maybe consider it until the "real" Fullscreen is supported with it?

**Comment by ivan-mogilko** [comment]
Alright. Proper fixing everything will take a while anyway.

But I said it wrong, the multitask fix #1292 is not simply "not enough", it actually cannot be applied to sdl2 version at all, as it will halt the update thread endlessly. Anyway, there seem to be a number of serious issues, so I better get to work on them asap.

**Comment by ivan-mogilko** [comment]
I actually managed to adjust a fix for 3.6.0 with a minimal addition (81aa5509e929442ef8ba6eadf3dccaf3e0e7b818), for some reason this idea did not come to me yesterday.

That takes care of suspending game updates while there's no input focus, but I suspect that may be not enough for Direct3D fullscreen situation. BTW, allegro engine version had a mysterious "Delay(1000)" when handling this event, now i think that maybe this was made to make sure that game update is halted before hiding the window... just a guess.
If Direct3D still crashes/hangs there has to be an additional rendering prevention before window gets minimized.

**Comment by ericoporto** [comment]
just for a note, with the temporary fix now merged I don't have this problem anymore in Win10.

**Comment by PaddyMac** [comment]
I am not entirely sure if this is the same issue I have been experiencing, but it certainly sounds the same. using 3.5.1.12 on Linux and using OpenGL, I also experience hanging of AGS when Alt+Tabing away from the game in fullscreen mode. It seems to be fine in windowed mode. Just Alt+Tabing away doesn't always immediately hang AGS, but if I try to start recording the AGS game with OBS, then it definitely hangs. The hard drive light on my PC also turns on and stays on until I force close AGS.

**Comment by ericoporto** [comment]
@PaddyMac you get this in any game or just a particular one?

**Comment by PaddyMac** [comment]
So far, the only game I've tried is The Cat Lady (GoG release). I only know of one other game in my library that I know uses AGS, so I'll give that one a try when I have the opportunity.

**Comment by ivan-mogilko** [comment]
I'm back to this problem, but would need to recall the details. IIRC since the multitasking mode has been fixed; and also a proper switch was introduced between "real" and "desktop" fullscreen mode. Need to double check that "real" fullscreen works correctly on Windows now, and then see what can be done to close this bug report.

@PaddyMac 
> I am not entirely sure if this is the same issue I have been experiencing, but it certainly sounds the same. using 3.5.1.12 on Linux and using OpenGL

The issue reported was experienced only with Direct3D exclusive fullscreen mode, for specific reasons. I don't remember anyone mentioning game hanging when using OpenGL before.
Unlike D3D, OpenGL is running always in windowed mode, its "fullscreen" is actually a fake (borderless window covering whole screen). It may or may not have similar problem when the window is switched off and minimized, but I would rather investigate it as a separate problem.

**Comment by ivan-mogilko** [comment]
In regards to "multitasking" mode, don't know how i missed this earlier, but i just realized that originally AGS prevented multitasking mode from working in fullscreen at all. This is both mentioned in [documentation](https://github.com/adventuregamestudio/ags-manual/wiki/Globalfunctions_General#setmultitaskingmode):
> Note that mode 1 does not work with some graphics cards in full-screen mode, so you should only rely on it working when your game is run in windowed mode.

and may be noticed in the code: 
https://github.com/adventuregamestudio/ags/blob/master/Engine/ac/global_game.cpp#L678-L680

The silly problem is that this was before engine supported switching between modes at runtime. When introducing the mode switch, i missed this miltitasking restriction, so now it's probably possible to start in windowed mode with multitasking mode on, and then switch to fullscreen, and this mode will be kept.

Tbh I suppose that it may be allowed in fullscreen so long as we may prevent renderer from failing / hanging the program, but it definitely should be brought to consistent behavior (either allowed in fs or not).

**Comment by ivan-mogilko** [comment]
In regards to hanging, another ridiculous thing which i somehow missed is, our engine code potentially puts the render function in an infinite loop here: https://github.com/adventuregamestudio/ags/blob/master/Engine/ac/draw.cpp#L840

It runs Present attempts indefnitely, hoping that it will eventually succeed. While doing so it will probably not poll system events, etc, so hypothetically may lock the rest of the program. This has to be fixed.

**Comment by ericoporto** [comment]
@ivan-mogilko  after 9f091926e393f75c90c951f9597710bf17395332 now the engine hangs for me when alt+tab a windows when using Direct3D renderer on Windows.

**Comment by ivan-mogilko** [comment]
> @ivan-mogilko after [9f09192](https://github.com/adventuregamestudio/ags/commit/9f091926e393f75c90c951f9597710bf17395332) now the engine hangs for me when alt+tab a fullscreen game when using Direct3D renderer on Windows.

Did not it hang for you before that? This is what the ticket was about.

**Comment by ivan-mogilko** [comment]
Ah! Eureka, I found why it's not resetting in fullscreen. The call to `IDirect3DDevice9::Reset` was failing with `D3DERR_INVALIDCALL` after switching back, and it took me a while debugging to realize why.

Apparently, when the program is switched out, it's window is reduced a little bit. This makes the engine to receive `SDL_WINDOWEVENT_SIZE_CHANGED`, and when it does it goes here:
https://github.com/adventuregamestudio/ags/blob/e8a1af3bdeb309c67d2cc120150fe5bf317783d5/Engine/main/engine.cpp#L1389-L1394

and then eventually here:
https://github.com/adventuregamestudio/ags/blob/e8a1af3bdeb309c67d2cc120150fe5bf317783d5/Engine/platform/windows/gfx/ali3dd3d.cpp#L776-L786

which effectively corrupts the fullscreen mode width & height parameters. Because when it tries to Reset device back, the mode is already different, and it simply cannot init the exclusive fullscreen with these.

So, the cause is clear now, and making a quick hackish workaround proved that restoring true parameters make it work. So the remaining issue is to code a clean solution for this.

**Comment by ivan-mogilko** [comment]
Pushed a fix, could you test again? 25329949f4d4a17ea8e6ad11fc7bd0ecccdfaca7
Btw, I tried SDL2's software renderer created with Direct3D driver, and it works fine, but maybe it always did.

> Is the event SDL_WINDOWEVENT_SIZE_CHANGED produced in both changes for you or only in some

We are logging this event, and it happens only when switching out; but then again it should not automatically resize back unless we restore a proper exclusive mode.

> I absolutely believe that the default Fullscreen should be bordless windows

IIRC borderless window is the default in the engine, but maybe not in the editor when it writes config, we may double check that.

**Comment by ericoporto** [comment]
It looks like this for me, when it goes full black is because it's on Exclusive Fullscreen and I can't use OBS in these cases.

https://user-images.githubusercontent.com/2244442/174903896-935b2f9c-9f0a-49ed-b3fd-0872dba835cb.mp4

As you can see, Old Skies used to hang (3.6.0.24) and with the latest engine from current master (2f04e3b81db598eb7950668ce9d88a145f796f2c), it still hangs. Old Skies has multitasking mode activated, it also has Direct3D renderer and Exclusive Fullscreen.

**Comment by ericoporto** [comment]
Dismiss my previous comment. It looks like the last commit indeed fixed!!!

**Comment by ivan-mogilko** [comment]
Note: multitasking mode is supposed to turn off automatically in exclusive fullscreen.

**Comment by ericoporto** [comment]
This still happens, it doesn't trigger right away for me, but if I have real fullscreen, and direct3D, and alt+tab a bunch of times, eventually AGS enters in a hanging state, in Win10.

**Comment by ericoporto** [comment]
I tried to add `game_update_suspend = false;` in `display_switch_in` and `display_switch_out`, and I also felt like it hang less easily. But still hangs. When hitting pause in the debugger, it's like it's always printing the debug error message, but I expected that, when it's hanged, hitting pause would hit inside the while that is meant to wait the game be not suspended anymore.

**Comment by ivan-mogilko** [comment]
So, this is because Reset never gets called after catching Ali3DFullscreenLostException, and it keeps endless loops, not being able to restore?

**Comment by ivan-mogilko** [comment]
`Ali3DFullscreenLostException` is getting thrown in case `TestCooperativeLevel` returns `D3DERR_DEVICELOST`. In such case Reset is never called; Reset is called only when `TestCooperativeLevel` returned `D3DERR_DEVICENOTRESET`. Comparing with the gamedev.net page you linked, this is almost equivalent to what their code example sais (unless I am missing something).

On Ali3DFullscreenLostException the engine prints "Renderer exception: %s" to the log. Do you see these messages when the engine hangs for you? I has not been able to reproduce this yet (on win7).

**Comment by ivan-mogilko** [comment]
It actually quite possible that the loop on line 934 is not needed, or is wrong at all. The idea was that it waits there until the game unsuspends, in case it got into the rendering while switching out. But maybe it's a wrong thing to do to wait there. Maybe the better way would be to instead break out of the external loop (one that starts at line 920) if `game_update_suspend` is detected.

While the game is suspended the `render_to_screen` is not supposed to be called anyway. So this situation may happen only if engine's main thread got there at the same time as game was switching out. If I remember things correctly.

**Comment by ivan-mogilko** [comment]
Hmm, comparing 3.6.0 with 3.5.1, the 3.5.1 called Delay under `catch (Ali3DFullscreenLostException) ` regardless of whether the game was suspended or not, which probably made more sense.

So, I think that regardless of this issue, the code of the internal loop could be changed from `while` to `do/while`, as in:

```
            do
            {
                sys_evt_process_pending();
                platform->Delay(300);
            }
           while (game_update_suspend && (!want_exit) && (!abort_engine));
```

**Comment by ericoporto** [comment]
> To clarify on hanging, are you telling that the engine hangs in this loop?

Logically, I think it should, but hitting "pause" in the debugger it's never there, it's always at some step of it printing the exception in the debugger, curiously if I kill the engine, at most 3 or 4 lines of such printing hit the console. So, not sure it hangs in that loop, I haven't been able to actually confirm it.

I wonder if I am configuring something wrongly in the debugger that makes exception be something different there.

> Are not they are handled by the call to sys_evt_process_pending?

Yes, I think so.

> be changed from while to do/while

I agree. It doesn't actually prevent the hanging though.

> Does this mean that it keeps getting D3DERR_DEVICELOST forever?

I *think* this is what I am observing.

**Comment by ericoporto** [comment]
I used the AGS build from [here](https://cirrus-ci.com/task/6275534488338432), ran Quest For Glory 2, and the log quickly gets MBs of `Renderer exception: Direct3D device is lost`. I cut those lines at the end.

<details><summary>here's a log where I managed to hang in the very few alt+tab</summary>
Adventure Game Studio v3.6 Interpreter
Copyright (c) 1999-2011 Chris Jones and 2011-2022 others
ACI version 3.6.0.36

Installing exception handler
Initializing backend libs
SDL Version: 2.24.1
Initializing game data
Looking for the game data.
 Cwd: C:/Users/erico/Documents/AGSProjects/Quest for Glory II
 Path arg: 
Searching in (cwd): C:/Users/erico/Documents/AGSProjects/Quest for Glory II
Found game config: C:/Users/erico/Documents/AGSProjects/Quest for Glory II/acsetup.cfg
 Cfg: data file: 
Searching for game data in: C:/Users/erico/Documents/AGSProjects/Quest for Glory II
Found game data pak: C:/Users/erico/Documents/AGSProjects/Quest for Glory II/Qfg2vga.exe
Located game data pak: c:/Users/erico/Documents/AGSProjects/Quest for Glory II/Qfg2vga.exe
Opened game data file: game28.dta
Game data version: 42
Compiled with: 3.2.0
Startup directory: C:/Users/erico/Documents/AGSProjects/Quest for Glory II
Data directory: c:/Users/erico/Documents/AGSProjects/Quest for Glory II/
Setting up game configuration
Unable to init voice pack 'speech.vox', file not found or of unknown format.
audio.vox found and initialized.
Initializing TTF renderer
Initializing mouse: number of buttons reported is 3
Install timer
Initializing audio
Requested audio driver: default
Audio driver: wasapi
AudioCore: opened device "Default OpenAL playback device"
Supported sound decoders:
 - MIDI decoder, using a subset of TiMidity : MIDI,MID,
 - Play modules through ModPlug : 669,AMF,AMS,DBM,DMF,DSM,FAR,GDM,IT,MDL,MED,MOD,MT2,MTM,OKT,PTM,PSM,S3M,STM,ULT,UMX,XM,
 - MPEG-1 Audio Layer I-III : MP3,MP2,MP1,
 - Microsoft WAVE audio format : WAV,
 - Audio Interchange File Format : AIFF,AIF,
 - Sun/NeXT audio file format : AU,
 - Ogg Vorbis audio : OGG,
 - Creative Labs Voice format : VOC,
 - Raw audio : RAW,
 - Shorten-compressed audio data : SHN,
 - Free Lossless Audio Codec : FLAC,FLA,
Install exit handler
Initialize legacy path finder library
Load game data
Game title: 'Quest for Glory II'
Game uid (old format): `1027255500`
Game guid: '{a46a9171-f6f9-456c-9b2b-a509b560ddc0}'
Game GUI version: 115
Checking for disk space
Initializing resolution settings
Game native resolution: 320 x 200 (32 bit)
Mouse cursor graphic area: (0,0)-(319,199) (320x200)
Device display resolution: 2560 x 1440
Graphic settings: driver: D3D9, windowed: no, screen size: 2560 x 1440, game scale: proportional
Built library path: d3d9.dll
Using graphics factory: D3D9
Created graphics driver: Direct3D 9
Supported gfx modes (32-bit): 
	640x480;640x480;640x480;720x480;720x480;720x576;800x600;800x600;
	1024x768;1024x768;1152x864;1176x664;1176x664;1176x664;1280x720;1280x720;
	1280x720;1280x768;1280x800;1280x960;1280x960;1280x1024;1280x1024;1360x768;
	1360x768;1366x768;1366x768;1600x900;1600x900;1600x1024;1600x1024;1600x1200;
	1680x1050;1680x1050;1920x1080;1920x1080;1920x1080;1920x1080;1920x1080;1920x1200;
	1920x1440;1920x1440;2560x1440;2560x1440;2048x1080;2048x1080;
Attempting to find nearest supported resolution for screen size 2560 x 1440 (32-bit) fullscreen
Attempt to switch gfx mode to 2560 x 1440 (32-bit) fullscreen
Graphics mode set: 2560 x 1440 (32-bit) fullscreen
Render frame set, render dest (128, 0, 2431, 1439 : 2304 x 1440)
Requested gfx filter: StdScale
Graphics filter set: 'StdScale', filter dest (128, 0, 2431, 1439 : 2304 x 1440)
Using gfx filter: StdScale
Mouse speed control: disabled, unit: 1.000000, user value: 1.000000
Mouse cursor graphic area: (128,0)-(2431,1439) (2304x1440)
Multitasking mode set: 0
Setting up window
Multitasking mode set: 0
Displaying preload image
Initialize sprites
Initialize game settings
Prepare to start game
Engine initialization complete
Starting game
SetMultitasking: overridden by fullscreen: 1 -> 0
Multitasking mode set: 0
Window event: size changed (2560, 1440)
Render frame set, render dest (128, 0, 2431, 1439 : 2304 x 1440)
Mouse cursor graphic area: (128,0)-(2431,1439) (2304x1440)
Window event: focus lost
Suspending the game on switch out
Switching out from the game
Renderer exception: Direct3D device is lost
Window event: focus gained
Resuming the game on switch in
Switching back into the game
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
Renderer exception: Direct3D device is lost
</details>

Here's a different log where it took me a little more time [qfg2vga_alt_tab_bug_log.zip](https://github.com/adventuregamestudio/ags/files/9856200/qfg2vga_alt_tab_bug_log.zip)

**Comment by ivan-mogilko** [comment]
Hm, would it work if you try 3.5.1 on same system?

**Comment by ericoporto** [comment]
Erh, apparently no, it hangs in a different way that it captures my mouse forever and I can't kill AGS and need to restart my machine. Used the latest release of AGS 3.5.1.

**Comment by ivan-mogilko** [comment]
For the reference this is how it's handled in SDL2:
https://github.com/libsdl-org/SDL/blob/053b5f85f164dd2f730a6500096a666001dc9d5b/src/render/direct3d/SDL_render_d3d.c#L1391

Does the 3.6.0 sdl2's software renderer work if it's forced to be direct3d in config?

**Comment by ericoporto** [comment]
No, software driver with direct3d never hangs!

**Comment by ivan-mogilko** [comment]
Well, these logs don't show much tbh, except it's clear that our renderer gets stuck in D3DERR_DEVICELOST for some obscure reason. I hope to find time and compare with SDL2 more closely tomorrow, but a simple hypothesis is that either we do something that we should not, or not do something that is necessary after this is detected.

**Comment by ericoporto** [comment]
Gave it a shot at adding more log. As I mentioned previously, with d3d not all alt+tabs are hanging, it usually takes some tries to cause a hanging. In the log, the alt+tabs that doesn't hang have a `Window event: size changed (2560, 1440)` happening, and once it hangs, this event doesn't happen. This occurs right after `display_switch_in_resume: game_update_suspend = false`.

- Added these log messages: https://github.com/ericoporto/ags/commit/3208495653d24c8bc07a44b4284b87788c4027a2
- Log for the software direct3d: [gist.github.com/ericoporto/642b80e5826301666a8b546989924859](https://gist.github.com/ericoporto/642b80e5826301666a8b546989924859)
- Log for hanging d3d renderer: [gist.github.com/ericoporto/dd954a7539cadc2d4db68b3374657688](https://gist.github.com/ericoporto/dd954a7539cadc2d4db68b3374657688)

**Comment by ivan-mogilko** [comment]
@ericoporto my tests may be not accurate, or this depends on some factors, but the change I mentioned earlier actually fixes the problem for me on win7. I pushed it as 27306da4cd0c1fa193f984644aa1915c8a944364. Previously you mentioned that it does not fix the problem for you though, but could you test just in case?

**Comment by ericoporto** [comment]
@ivan-mogilko , this change does make it harder for me hit the bug, but I can still cause the situation to happen through alt+tabbing a bit quicker

**Comment by ivan-mogilko** [comment]
Hm, I cannot reproduce this no matter how much I try.

Maybe it depends on delay value? In the previous version the delay was 500 ms.

**Comment by ericoporto** [comment]
> Maybe it depends on delay value?

Tried 100ms and 1000ms, doesn't change much regards to hanging, but the smaller value reduced the blinking when alt+tab back in - when it works and doesn't hang.

> does this happen in any game, or in particular game / during particular scene?

Any game. I can reproduce it even by just loading an empty sierra template game or the one made for the GUI issue.

**Comment by ivan-mogilko** [comment]
So, something I noticed reading your log from above: this is how the switching into the game normally looks like:
```
Window event: focus gained
Resuming the game on switch in
Switching back into the game
display_switch_in_resume: video_resume()
display_switch_in_resume: game_update_suspend = false
Window event: size changed (2560, 1440)
Render frame set, render dest (128, 0, 2431, 1439 : 2304 x 1440)
Mouse cursor graphic area: (128,0)-(2431,1439) (2304x1440)
render_to_screen: gfxDriver->Render()
ResetDeviceIfNecessary: -2005530519 = direct3ddevice->TestCooperativeLevel()
ResetDeviceIfNecessary: after 0 = ResetD3DDevice()
```

This repeats 4 times, but the 5th last time looks like this:
```
Window event: focus gained
Resuming the game on switch in
Switching back into the game
display_switch_in_resume: video_resume()
display_switch_in_resume: game_update_suspend = false
render_to_screen: gfxDriver->Render()
ResetDeviceIfNecessary: -2005530520 = direct3ddevice->TestCooperativeLevel()
ResetDeviceIfNecessary: throw Ali3DFullscreenLostException()
```

To be precise, following part is missing in the case which leads to the hanging:
```
Window event: size changed (2560, 1440)
```

**Comment by ericoporto** [comment]
https://user-images.githubusercontent.com/2244442/199126626-203ce4f6-cf54-455e-ae4d-c84ee4a6e0c1.mp4

So just to illustrate how the hanging manifests here, made a video, once the hang happens, the mouse cursor disappears - not sure if the mouse cursor disappearing here is relevant. I need to use Windows Key+Tab to move AGS to a new workspace before I can go back to the previous workspace and use the task manager to kill AGS. 

Screen recording misses some frames, when alt+tab happens, so I had to film the screen.

**Comment by ericoporto** [comment]
Recently Manu in the forums reported a similar problem, but it requires a GUI to reproduce.

https://www.adventuregamestudio.co.uk/forums/ags-engine-editor-releases/ags-3-6-0-release-candidate-rc1/msg636652983/#msg636652983

[CrashInfo.3.6.0.41.zip](https://github.com/adventuregamestudio/ags/files/10527238/CrashInfo.3.6.0.41.zip)

**Comment by ivan-mogilko** [comment]
That recent crash should be fixed by 278ba3b3d49618e91c09f60ab93b36ae0fef0d28.

But there's still problem that IDirect3DDevice9::Reset fails with code 0x8876086C sometimes (like each second time).

**Comment by ivan-mogilko** [comment]
So, apparently, engine also has to react to "switch in" event and release render targets in case we're in real fullscreen mode.
Pushed 1c337a31ca81b8d738a3dc6023363aa3c996eb61 (also see comments to commit).

**Comment by ericoporto** [comment]
Here's a log of a hang: [log_hang.txt.zip](https://github.com/adventuregamestudio/ags/files/10609519/log_hang.txt.zip)
I added extra log messages as here: https://github.com/ericoporto/ags/commit/b11d6ac072e49815f1aeafe3bbe6fdf5a2af626f

**Comment by ivan-mogilko** [comment]
> Here's a log of a hang: [log_hang.txt.zip](https://github.com/adventuregamestudio/ags/files/10609519/log_hang.txt.zip) I added extra log messages as here: [ericoporto@b11d6ac](https://github.com/ericoporto/ags/commit/b11d6ac072e49815f1aeafe3bbe6fdf5a2af626f)
> 
> And here is the same hang, but using the AGS 3.6.0.42 release, without the additional debug messages: [log_hang2.txt.zip](https://github.com/adventuregamestudio/ags/files/10609521/log_hang2.txt.zip)

I have one small request, could you please print HRESULT as hex `%x`, that makes it easier to lookup the error codes.

Anyways, if my understanding is correct, the last HRESULT returned by the TestCooperativeLevel is D3DERR_INVALIDCALL.

I do not know why it does that, but the immediate problem is that our code in ResetDeviceIfNecessary only tests for two errors: D3DERR_DEVICELOST and D3DERR_DEVICENOTRESET. This means that after D3DERR_INVALIDCALL the renderer continues to execute instead of bailing out. Probably, this leads to more errors, or hang up.

**Comment by ericoporto** [comment]
Here's the log using %x: [log.txt.zip](https://github.com/adventuregamestudio/ags/files/10609860/log.txt.zip)
And the change that added the log messages: https://github.com/ericoporto/ags/commit/1aaeef96bf03222e95d26c9880b8c49c5f7c85a2

After extra logs: [log_v2_ResetD3DDevice.txt.zip](https://github.com/adventuregamestudio/ags/files/10609876/log_v2_ResetD3DDevice.txt.zip)
Here https://github.com/ericoporto/ags/commit/313bc3b51c01c94ce9579183d1dc41a6812df3ff

**Comment by ericoporto** [comment]
I was thinking about this, would it make sense to add an `[disabled]` entry for `fullscreen` so one can lock it to `full_window`, essentially making it the only way to run in FullScreen? I think in SDL3 there isn't the exclusive fullscreen mode anymore, so this problem will eventually take care of itself in the future.

**Comment by ivan-mogilko** [comment]
I don't know if it matters for this issue, but just now I found that potentially there could be render surface pointers not restored after device reset. I only noticed this problem in the current master branch, after PR #2353, where I replaced raw surface pointers with smart ptrs. Since this seem like a legit bug, I backported the fix to release-3.6.0, e.g. see a4fb5e379a829be572c75021ed34cffc9bf69e29

**Comment by ericoporto** [comment]
I don't have the issue anymore but I also have recently upgraded my Nvidia Drivers that I had not upgraded in two years... I remember before upgrading that I didn't had the issue anymore in 3.6.1 but had in 3.6.0 at the time. I tried previous versions of 3.6.0 and can't reproduce anymore.... So whatever issue that had, it got fixed some time ago, but the one I still had was an issue on my particular computer.
